File: src/python/indexer.py
"""indexer.py — procedural filepath indexer, one filetype at a time.

Design mirrors claude-science audit/bench_cpu.py: trial-safe, JSON-serializable,
fresh-process friendly, with explicit --json provenance.

Usage:
  from src.python.indexer import index_one_type, index_many_types, iter_types

  idx = index_one_type("/home/zhouk/.claude-science", ".json")
  print(idx.count, idx.total_bytes)
  for entry in idx.entries:
      print(entry.relative_path, entry.parts.subdirs, entry.parts.filename)

  # one type at a time loop (procedural):
  for ft, idx in iter_types(root, [".py",".json",".md"]):
      ...

CLI:
  python -m src.python.indexer /tmp --ext .py --json out.json
  python -m src.python.indexer /tmp --ext .py --all --json out.json  # indexes all distinct exts
"""
from __future__ import annotations
import argparse
import json
import os
import time
from pathlib import Path
from typing import Iterator, List, Optional, Set

from .models import FileEntry, FileTypeIndex, PathParts, IndexResult


# same set as in audit-compute-env.sh for ignore discipline
DEFAULT_IGNORE_DIRS: Set[str] = {
    ".git", ".hg", ".svn", "__pycache__", ".venv", "venv", ".mypy_cache",
    ".pytest_cache", "node_modules", ".next", ".turbo", "dist", "build",
    ".parcel-cache", ".claude", ".codex",  # but caller can override to include .claude-science
}
DEFAULT_IGNORE_FILES: Set[str] = set()


def _should_ignore_dir(name: str, ignore_dirs: Set[str]) -> bool:
    return name in ignore_dirs


def _normalize_ext(ext: str) -> str:
    """Ensure leading dot, lowercased. '.JSON' -> '.json', 'json' -> '.json', '' -> ''."""
    if not ext:
        return ""
    e = ext.strip().lower()
    if not e.startswith("."):
        e = "." + e
    return e


def iter_files_recursive(
    root: Path,
    ext: Optional[str],
    ignore_dirs: Set[str] = DEFAULT_IGNORE_DIRS,
    follow_symlinks: bool = False,
) -> Iterator[Path]:
    """Yield absolute Path objects recursively, filtered to one ext if given.

    Procedural core: uses os.scandir iteratively (not glob) so subdirectory
    parts are available without extra stat calls. One filetype at a time
    keeps memory bounded: caller processes .py, then .json, etc.
    """
    root = root.resolve()
    stack: List[Path] = [root]
    while stack:
        cur = stack.pop()
        try:
            with os.scandir(cur) as it:
                for entry in it:
                    try:
                        if entry.is_dir(follow_symlinks=follow_symlinks):
                            if _should_ignore_dir(entry.name, ignore_dirs):
                                continue
                            stack.append(Path(entry.path))
                        elif entry.is_file(follow_symlinks=follow_symlinks):
                            p = Path(entry.path)
                            if ext is None:
                                yield p
                            else:
                                # Path.suffix is lowercased via _normalize
                                if p.suffix.lower() == ext:
                                    yield p
                    except (PermissionError, OSError):
                        continue
        except (PermissionError, OSError, FileNotFoundError):
            continue


def index_one_type(
    root: str | Path,
    file_type: str,
    ignore_dirs: Optional[Set[str]] = None,
    follow_symlinks: bool = False,
    compute_hash: bool = False,
    sort_by: str = "relative_path",
) -> FileTypeIndex:
    """Procedurally index ONE filetype under root.

    Returns FileTypeIndex with decomposed PathParts for every match.
    Focuses on one ext so caller can bound VRAM/RAM and stream types
    sequentially (like bench_cpu sweeps one thread count at a time).
    """
    t0 = time.perf_counter()
    root_p = Path(root).resolve()
    if not root_p.exists():
        raise FileNotFoundError(f"root does not exist: {root_p}")
    if not root_p.is_dir():
        raise NotADirectoryError(f"root is not a directory: {root_p}")

    ext = _normalize_ext(file_type)
    # allow caller to index dotless hidden files like "" by passing ""
    # but default is to require ext
    if ignore_dirs is None:
        ignore_dirs = DEFAULT_IGNORE_DIRS

    entries: List[FileEntry] = []
    scanned_dirs = 0
    total_bytes = 0

    # count scanned dirs separately for provenance
    # we walk again with os.walk for dir count when needed, but cheaper:
    # increment on each stack pop inside iter_files_recursive — we inline here
    # for single-pass efficiency.
    stack: List[Path] = [root_p]
    seen_files: List[Path] = []  # collected then stat'd — keeps scandir fast

    # Single walk that both counts dirs and collects matching files
    while stack:
        cur = stack.pop()
        scanned_dirs += 1
        try:
            with os.scandir(cur) as it:
                for entry in it:
                    try:
                        if entry.is_dir(follow_symlinks=follow_symlinks):
                            if _should_ignore_dir(entry.name, ignore_dirs):
                                continue
                            stack.append(Path(entry.path))
                        elif entry.is_file(follow_symlinks=follow_symlinks):
                            p = Path(entry.path)
                            if p.suffix.lower() == ext:
                                seen_files.append(p)
                    except (PermissionError, OSError):
                        continue
        except (PermissionError, OSError, FileNotFoundError):
            continue

    # Now stat and build FileEntry objects — still procedural, one file at a time
    for abs_path in seen_files:
        try:
            stat = abs_path.stat()
            rel = abs_path.relative_to(root_p)
            parts = PathParts.from_relative(rel)
            sha = None
            if compute_hash:
                import hashlib
                h = hashlib.sha256()
                with open(abs_path, "rb") as f:
                    for chunk in iter(lambda: f.read(1 << 20), b""):
                        h.update(chunk)
                sha = h.hexdigest()
            fe = FileEntry(
                relative_path=str(rel).replace(os.sep, "/"),  # posix for portability
                absolute_path=str(abs_path),
                parts=parts,
                file_type=ext,
                size_bytes=stat.st_size,
                mtime_ns=stat.st_mtime_ns,
                depth=parts.depth,
                sha256=sha,
            )
            entries.append(fe)
            total_bytes += stat.st_size
        except (OSError, ValueError):
            continue

    # deterministic order — critical for diffing / caching like bench_cpu trials
    if sort_by == "relative_path":
        entries.sort(key=lambda e: e.relative_path)
    elif sort_by == "size":
        entries.sort(key=lambda e: e.size_bytes, reverse=True)
    elif sort_by == "mtime":
        entries.sort(key=lambda e: e.mtime_ns, reverse=True)

    elapsed_ms = (time.perf_counter() - t0) * 1000
    return FileTypeIndex(
        root=str(root_p),
        file_type=ext,
        count=len(entries),
        total_bytes=total_bytes,
        entries=entries,
        scanned_dirs=scanned_dirs,
        elapsed_ms=elapsed_ms,
    )


def iter_types(
    root: str | Path,
    file_types: List[str],
    **kwargs,
) -> Iterator[tuple[str, FileTypeIndex]]:
    """Procedural iterator: one filetype at a time, yielding (ext, index)."""
    for ft in file_types:
        ext = _normalize_ext(ft)
        yield ext, index_one_type(root, ext, **kwargs)


def discover_file_types(
    root: str | Path,
    ignore_dirs: Optional[Set[str]] = None,
    limit: Optional[int] = None,
) -> List[str]:
    """Scan once to discover distinct exts present — then caller can iterate one by one."""
    root_p = Path(root).resolve()
    if ignore_dirs is None:
        ignore_dirs = DEFAULT_IGNORE_DIRS
    exts: Set[str] = set()
    stack = [root_p]
    while stack:
        cur = stack.pop()
        try:
            with os.scandir(cur) as it:
                for entry in it:
                    try:
                        if entry.is_dir():
                            if entry.name in ignore_dirs:
                                continue
                            stack.append(Path(entry.path))
                        elif entry.is_file():
                            ext = Path(entry.name).suffix.lower()
                            # include "" for extensionless files
                            exts.add(ext)
                            if limit and len(exts) >= limit:
                                return sorted(exts)
                    except OSError:
                        continue
        except OSError:
            continue
    return sorted(exts)


def index_many_types(
    root: str | Path,
    file_types: Optional[List[str]] = None,
    ignore_dirs: Optional[Set[str]] = None,
    **kwargs,
) -> IndexResult:
    """Index several types sequentially — still one at a time internally."""
    t0 = time.perf_counter()
    root_p = Path(root).resolve()
    if file_types is None:
        file_types = discover_file_types(root_p, ignore_dirs=ignore_dirs)
    indexes: List[FileTypeIndex] = []
    for ext in file_types:
        idx = index_one_type(root_p, ext, ignore_dirs=ignore_dirs, **kwargs)
        indexes.append(idx)
    elapsed_ms = (time.perf_counter() - t0) * 1000
    return IndexResult(root=str(root_p), indexes=indexes, elapsed_ms=elapsed_ms)


# --- CLI mirroring audit/audit-compute-env.sh style ---
def main() -> None:
    ap = argparse.ArgumentParser(description="Procedural filepath indexer — one filetype at a time")
    ap.add_argument("root", help="root directory to walk")
    ap.add_argument("--ext", default=".py", help="single filetype to index, e.g. .json (default .py)")
    ap.add_argument("--all", action="store_true", help="discover and index all exts under root")
    ap.add_argument("--json", dest="json_out", help="write JSON to file (else stdout)")
    ap.add_argument("--ignore", nargs="*", help="extra dir names to ignore")
    ap.add_argument("--include-hidden", action="store_true", help="do not ignore .git/.venv/node_modules")
    ap.add_argument("--hash", action="store_true", help="compute sha256 per file")
    ap.add_argument("--sort", choices=["relative_path", "size", "mtime"], default="relative_path")
    ap.add_argument("--limit", type=int, help="limit entries printed to stdout (does not affect JSON)")
    args = ap.parse_args()

    ignore = set(DEFAULT_IGNORE_DIRS) if not args.include_hidden else set()
    if args.ignore:
        ignore.update(args.ignore)

    if args.all:
        res = index_many_types(args.root, ignore_dirs=ignore, compute_hash=args.hash, sort_by=args.sort)
        out = json.dumps(res.to_dict(), indent=2)
        print(f"Indexed {len(res.indexes)} types under {res.root} in {res.elapsed_ms:.1f}ms")
        for ix in res.indexes:
            print(f"  {ix.file_type or '(no ext)':12s} {ix.count:6d} files  {ix.total_bytes/1024:.1f} KiB  {ix.scanned_dirs} dirs")
    else:
        idx = index_one_type(args.root, args.ext, ignore_dirs=ignore, compute_hash=args.hash, sort_by=args.sort)
        print(f"Indexed {idx.count} *{idx.file_type} files under {idx.root} ({idx.total_bytes} bytes, {idx.scanned_dirs} dirs, {idx.elapsed_ms:.1f}ms)")
        for e in idx.entries[: args.limit or 20]:
            print(f"  {e.relative_path:60s}  {e.parts.subdirs}  {e.size_bytes}B")
        if args.limit and idx.count > args.limit:
            print(f"  ... +{idx.count - args.limit} more")
        out = idx.to_json()

    if args.json_out:
        Path(args.json_out).write_text(out)
        print(f"Wrote {args.json_out}")
    elif args.all:
        # already printed summary; dump full JSON to stdout when no --json
        pass


if __name__ == "__main__":
    main()

File: src/typescript/indexer.ts
/**
 * indexer.ts — procedural filepath indexer, one filetype at a time.
 * Mirrors Python src/python/indexer.py exactly (same ignore rules, same PathParts).
 * Uses zod schemas from schemas.ts to validate each FileEntry.
 *
 * Usage:
 *   npx tsx src/typescript/indexer.ts /home/zhouk/.claude-science --ext .json
 *   npx tsx src/typescript/indexer.ts /home/zhouk/.claude-science --all
 */
import fs from "node:fs";
import path from "node:path";
import { fileEntrySchema, fileTypeIndexSchema } from "./schemas.js";
import type { FileEntry, FileTypeIndex, IndexResult, PathParts } from "./types.js";

const DEFAULT_IGNORE_DIRS = new Set([
  ".git",".hg",".svn","__pycache__",".venv","venv",".mypy_cache",
  ".pytest_cache","node_modules",".next",".turbo","dist","build",
  ".parcel-cache",".claude",".codex",
]);

function normalizeExt(ext: string): string {
  if (!ext) return "";
  let e = ext.trim().toLowerCase();
  if (!e.startsWith(".")) e = "." + e;
  return e;
}

function pathPartsFromRelative(relative: string): PathParts {
  // posix handling — relative is posix from Python side
  if (!relative || relative === ".") {
    return { subdirs: [], depth: 0, filename: "", stem: "", ext: "", parentRel: "", relativePath: "" };
  }
  const parts = relative.split("/").filter(Boolean);
  const filename = parts[parts.length - 1] ?? "";
  const subdirs = parts.slice(0, -1);
  const dot = filename.lastIndexOf(".");
  // hidden dotfile with no ext: ".legacy-adopted" -> stem=filename, ext=""
  // ".key-backups" is a dir, not a file, but for files we handle:
  let stem = filename;
  let ext = "";
  if (dot > 0) { // >0 so ".gitignore" stays stem=".gitignore" ext=""
    stem = filename.slice(0, dot);
    ext = filename.slice(dot).toLowerCase();
  } else if (filename.startsWith(".") && dot === -1) {
    stem = filename; ext = "";
  } else if (dot === -1) {
    stem = filename; ext = "";
  }
  const parentRel = subdirs.join("/");
  return { subdirs, depth: subdirs.length, filename, stem, ext, parentRel, relativePath: relative };
}

export function indexOneType(
  root: string,
  fileType: string,
  opts: { ignoreDirs?: Set<string>; followSymlinks?: boolean; sortBy?: "relativePath"|"size"|"mtime" } = {}
): FileTypeIndex {
  const t0 = performance.now();
  const rootAbs = path.resolve(root);
  if (!fs.existsSync(rootAbs) || !fs.statSync(rootAbs).isDirectory()) {
    throw new Error(`root does not exist or not a directory: ${rootAbs}`);
  }
  const ext = normalizeExt(fileType);
  const ignoreDirs = opts.ignoreDirs ?? DEFAULT_IGNORE_DIRS;
  const sortBy = opts.sortBy ?? "relativePath";

  const entries: FileEntry[] = [];
  let scannedDirs = 0;
  let totalBytes = 0;

  const stack: string[] = [rootAbs];
  const seen: string[] = [];

  while (stack.length) {
    const cur = stack.pop()!;
    scannedDirs++;
    let dirents: fs.Dirent[];
    try {
      dirents = fs.readdirSync(cur, { withFileTypes: true });
    } catch { continue; }
    for (const d of dirents) {
      const full = path.join(cur, d.name);
      try {
        if (d.isDirectory()) {
          if (ignoreDirs.has(d.name)) continue;
          stack.push(full);
        } else if (d.isFile() || (opts.followSymlinks && d.isSymbolicLink())) {
          const effExt = path.extname(d.name).toLowerCase();
          if (effExt === ext) seen.push(full);
        }
      } catch { continue; }
    }
  }

  for (const absPath of seen) {
    try {
      const stat = fs.statSync(absPath);
      const relPosix = path.relative(rootAbs, absPath).split(path.sep).join("/");
      const parts = pathPartsFromRelative(relPosix);
      const entry: FileEntry = {
        relativePath: relPosix,
        absolutePath: absPath,
        parts,
        fileType: ext,
        sizeBytes: stat.size,
        mtimeNs: Number(stat.mtimeMs * 1e6),
        depth: parts.depth,
        sha256: null,
      };
      // validate via zod — mirrors Python dataclass frozen guarantees
      fileEntrySchema.parse(entry);
      entries.push(entry);
      totalBytes += stat.size;
    } catch { continue; }
  }

  if (sortBy === "relativePath") entries.sort((a,b)=> a.relativePath.localeCompare(b.relativePath));
  else if (sortBy === "size") entries.sort((a,b)=> b.sizeBytes - a.sizeBytes);
  else if (sortBy === "mtime") entries.sort((a,b)=> b.mtimeNs - a.mtimeNs);

  const elapsedMs = performance.now() - t0;
  const idx: FileTypeIndex = {
    root: rootAbs,
    fileType: ext,
    count: entries.length,
    totalBytes,
    entries,
    scannedDirs,
    elapsedMs,
    indexedAt: new Date().toISOString(),
  };
  fileTypeIndexSchema.parse(idx);
  return idx;
}

export function discoverFileTypes(root: string, ignoreDirs: Set<string> = DEFAULT_IGNORE_DIRS): string[] {
  const rootAbs = path.resolve(root);
  const exts = new Set<string>();
  const stack = [rootAbs];
  while (stack.length) {
    const cur = stack.pop()!;
    let dirents: fs.Dirent[];
    try { dirents = fs.readdirSync(cur, { withFileTypes: true }); } catch { continue; }
    for (const d of dirents) {
      const full = path.join(cur, d.name);
      try {
        if (d.isDirectory()) {
          if (ignoreDirs.has(d.name)) continue;
          stack.push(full);
        } else if (d.isFile()) {
          exts.add(path.extname(d.name).toLowerCase());
        }
      } catch { continue; }
    }
  }
  return [...exts].sort();
}

export function indexManyTypes(root: string, fileTypes?: string[], ignoreDirs?: Set<string>): IndexResult {
  const t0 = performance.now();
  const types = fileTypes ?? discoverFileTypes(root, ignoreDirs);
  const indexes = types.map(ft => indexOneType(root, ft, { ignoreDirs }));
  return { root: path.resolve(root), indexes, elapsedMs: performance.now() - t0 };
}

// CLI mirroring Python
if (import.meta.url === `file://${process.argv[1]}`) {
  const root = process.argv[2] ?? "/home/zhouk/.claude-science";
  const extIdx = process.argv.indexOf("--ext");
  const ext = extIdx !== -1 ? process.argv[extIdx+1] : ".json";
  const all = process.argv.includes("--all");
  const jsonOutIdx = process.argv.indexOf("--json");
  const jsonOut = jsonOutIdx !== -1 ? process.argv[jsonOutIdx+1] : null;
  const includeHidden = process.argv.includes("--include-hidden");

  const ignore = includeHidden ? new Set<string>() : DEFAULT_IGNORE_DIRS;
  if (all) {
    const res = indexManyTypes(root, undefined, ignore);
    console.log(`Indexed ${res.indexes.length} types under ${res.root} in ${res.elapsedMs.toFixed(1)}ms`);
    for (const ix of res.indexes) console.log(`  ${ix.fileType || "(no ext)"} ${ix.count} files`);
    if (jsonOut) fs.writeFileSync(jsonOut, JSON.stringify(res, null, 2));
  } else {
    const idx = indexOneType(root, ext, { ignoreDirs: ignore });
    console.log(`Indexed ${idx.count} *${idx.fileType} files under ${idx.root} (${idx.totalBytes} bytes, ${idx.scannedDirs} dirs, ${idx.elapsedMs.toFixed(1)}ms)`);
    for (const e of idx.entries.slice(0,20)) console.log(`  ${e.relativePath}  ${JSON.stringify(e.parts.subdirs)}  ${e.sizeBytes}B`);
    if (jsonOut) fs.writeFileSync(jsonOut, JSON.stringify(idx, null, 2));
  }
}
